File: samples/gui/src/glwidget.cpp
#include <iostream>
#include <QPainter>
#include <QImage>

#include "glwidget.h"

GLWidget::GLWidget()
{

}

GLWidget::~GLWidget()
{

}

QRect GLWidget::getImageRect(const QImage& img) const
{
    float ratio = std::min((float)width() / (float)img.width(), (float)height() / (float)img.height());
    float w = (float)img.width() * ratio;
    float h = (float)img.height() * ratio;
    float x = ((float)width() - w) / 2.0f;
    float y = ((float)height() - h) / 2.0f;
    return QRect((int)x, (int)y, (int)w, (int)h);
}

void GLWidget::paintGL()
{
    if (f.isValid()) {
        mutex.lock();
        QPainter painter;
        painter.begin(this);
        QImage img = QImage(f.m_frame->data[0], f.m_frame->width, f.m_frame->height, QImage::Format_RGB888);
        painter.drawImage(getImageRect(img), img);
        painter.end();
        mutex.unlock();
    }
}

void GLWidget::renderCallback(const avio::Frame& frame)
{
    if (!frame.isValid()) {
        std:: cout << "render callback recvd invalid Frame" << std::endl;
        return;
    }
    mutex.lock();
    f = frame;
    mutex.unlock();
    update();
}

QSize GLWidget::sizeHint() const
{
    return QSize(640, 480);
}

File: samples/pyqt/glwidget.py
from PyQt6.QtOpenGLWidgets import QOpenGLWidget
from PyQt6.QtGui import QPainter, QImage
from PyQt6.QtCore import QRect, QSize
import numpy as np

class GLWidget(QOpenGLWidget):
    def __init__(self):
        super().__init__()
        self.image = QImage()
    
    def sizeHint(self):
        return QSize(640, 480)

    def renderCallback(self, F):
        ary = np.array(F, copy = False)
        h, w, d = ary.shape
        self.image = QImage(ary.data, w, h, d * w, QImage.Format.Format_RGB888)
        self.update()

    def getImageRect(self):
        ratio = min(self.width() / self.image.width(), self.height() / self.image.height())
        w = self.image.width() * ratio
        h = self.image.height() * ratio
        x = (self.width() - w) / 2
        y = (self.height() - h) / 2
        return QRect(int(x), int(y), int(w), int(h))

    def paintGL(self):
        if (not self.image.isNull()):
            painter = QPainter(self)
            painter.setRenderHint(QPainter.RenderHint.Antialiasing)
            painter.drawImage(self.getImageRect(), self.image)

    def clear(self):
        self.image.fill(0)
        self.update()
